proxy/pkg/zdmproxy/streamidmapper.go

```diff
@@ -0,0 +1,130 @@
+// hence they must have non-conflicting ids with user's requests.
+type StreamIdMapper interface {
+	GetNewIdFor(streamId int16) (int16, error)
+	RestoreId(syntheticId int16) (int16, error)
+	ReleaseId(syntheticId int16) (int16, error)
+}
+
+type cqlStreamIdMapper struct {
+	sync.Mutex
+	ids chan int16
+}
+
+func NewCqlStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	var ids = make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		ids <- i
+	}
+	return &cqlStreamIdMapper{
+		ids: ids,
+	}
+}
+
+func (icsim *cqlStreamIdMapper) GetNewIdFor(_ int16) (int16, error) {
+	icsim.Lock()
+	defer icsim.Unlock()
+	select {
+	case id, ok := <-icsim.ids:
+		if ok {
+			return id, nil
+		} else {
+			return -1, fmt.Errorf("stream id channel closed")
+		}
+	default:
+		return -1, fmt.Errorf("no stream id available")
+	}
+}
+
+func (icsim *cqlStreamIdMapper) RestoreId(_ int16) (int16, error) {
+	return -1, fmt.Errorf("internal CQL ids have no original ids to be restored")
+}
+
+func (icsim *cqlStreamIdMapper) ReleaseId(syntheticId int16) (int16, error) {
+	icsim.Lock()
+	defer icsim.Unlock()
+	select {
+	case icsim.ids <- syntheticId:
+	default:
+		log.Errorf("stream ids channel full, ignoring id %v", syntheticId)
+	}
+	return -1, nil
+}
+
+type streamIdMapper struct {
+	sync.Mutex
+	idMapper   map[int16]int16
+	synMapper  map[int16]int16
+	clusterIds chan int16
+}
+
+func NewStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	idMapper := make(map[int16]int16)
+	synMapper := make(map[int16]int16)
+	streamIdsQueue := make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		streamIdsQueue <- i
+	}
+	return &streamIdMapper{
+		idMapper:   idMapper,
+		clusterIds: streamIdsQueue,
+		synMapper:  synMapper,
+	}
+}
+
+func (sim *streamIdMapper) GetNewIdFor(streamId int16) (int16, error) {
+	sim.Lock()
+	defer sim.Unlock()
+	syntheticId, contains := sim.idMapper[streamId]
+	if contains {
+		return syntheticId, nil
+	}
+	select {
+	case id, ok := <-sim.clusterIds:
+		if ok {
+			syntheticId = id
+		} else {
+			return -1, fmt.Errorf("stream id channel closed")
+		}
+	default:
+		return -1, fmt.Errorf("no stream id available")
+	}
+	sim.idMapper[streamId] = syntheticId
+	sim.synMapper[syntheticId] = streamId
+	return syntheticId, nil
+}
+
+func (sim *streamIdMapper) RestoreId(syntheticId int16) (int16, error) {
+	sim.Lock()
+	defer sim.Unlock()
+	originalId, contains := sim.synMapper[syntheticId]
+	if contains {
+		return originalId, nil
+	}
+	return -1, fmt.Errorf("no matching id found for synthetic id %v", syntheticId)
+}
+
+func (sim *streamIdMapper) ReleaseId(syntheticId int16) (int16, error) {
+	sim.Lock()
+	defer sim.Unlock()
+	var originalId = sim.synMapper[syntheticId]
+	delete(sim.idMapper, originalId)
+	delete(sim.synMapper, syntheticId)
+	select {
+	case sim.clusterIds <- syntheticId:
+	default:
+		log.Errorf("stream ids channel full, ignoring id %v", syntheticId)
+	}
+	return originalId, nil
```

Comment:
I'm struggling to understand the difference between `ReleaseId` and `RestoreId` in practice. In the POV of the stream id mapper, it shouldn't matter if we are restoring or releasing, in practice the mapper needs to release the ID, delete it from the maps and return the original one. Whether this is being called because the request timed out or because a response was returned by the server shouldn't matter.